integrations/gemini/actions/response/pipeline/01_focusInput/main.ts

```diff
@@ -21,113 +22,208 @@ const systemPrompt = await Deno.readTextFile(
-      });
-    }
-    return;
-  }
-
-  const responseSchema = {
-    type: Type.OBJECT,
-    properties: inputRatingScales.reduce(
-      (properties, ratingScale) => {
-        properties[ratingScale.name] = {
-          type: Type.OBJECT,
-          properties: {
-            rating: {
-              type: Type.STRING,
-              description: ratingScale.definition,
-              enum: ["1", "2", "3", "4", "5"],
-            },
-            rationale: {
-              type: Type.STRING,
-              description: "A 1-2 sentence rationale for the rating given.",
-            },
-          },
-          required: ["rating"],
-        };
-
-        return properties;
-      },
-      {} as Record<string, Schema>,
-    ),
-    required: inputRatingScales.map((ratingScale) => ratingScale.name),
-  };
-
-  const geminiCalls: Promise<void>[] = [];
-  let pointer = 0;
-  while (pointer < input.length) {
-    const event = input[pointer];
-
-    if (event.lastProcessedAt) {
-      pointer++;
-      continue;
-    }
-
-    if (
-      event.user?.id === this.action.user?.id ||
-      this.action.service.system.isSystemUser(event.user) ||
-      [
-        BottEventType.ACTION_START,
-        BottEventType.ACTION_OUTPUT,
-        BottEventType.ACTION_COMPLETE,
-        BottEventType.ACTION_ABORT,
-      ].includes(event.type as BottEventType)
-    ) {
-      // We still want to see our own messages in the context window
-      // But we don't need to rate them
-      pointer++;
-      continue;
-    }
-
-    geminiCalls.push((async () => {
-      const currentPointer = pointer;
-      const scoresWithRationale = await generateFromEvents<
-        Record<string, { rating: string; rationale: string | undefined }>
-      >(
-        input.slice(0, currentPointer + 1),
-        {
-          systemPrompt,
-          responseSchema,
-          pipeline: this,
-          model: GEMINI_RATING_MODEL,
-          useThirdPersonAnalysis: true,
-        },
+  return await withSpanAsync(
+    "focusInput",
+    { inputEventCount: this.data.input.length },
+    async (span) => {
+      const input = this.data.input;
+      const inputReasons = this.action.service.system.settings.reasons.input;
+      const inputRatingScales = [
+        ...new Set(inputReasons.flatMap((reason) => reason.ratingScales ?? [])),
+      ];
+
+      span.setAttribute("inputReasons.count", inputReasons.length);
+      span.setAttribute(
+        "inputReasons.names",
+        inputReasons.map((r) => r.name).join(","),
+      );
+      span.setAttribute("inputRatingScales.count", inputRatingScales.length);
+      span.setAttribute(
+        "inputRatingScales.names",
+        inputRatingScales.map((s) => s.name).join(","),
       );
 
-      const ratings: Record<string, number> = {};
+      console.log(`[focusInput] Processing ${input.length} input events`);
+      console.log(
+        `[focusInput] Reasons: ${inputReasons.map((r) => r.name).join(", ")}`,
```

Comment:
let's log full objects and trust the provider to handle serialization (otel viewer allows for filtering/collapsing of items)